iina/LegacyMigration.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  LegacyMigration.swift
+//  iina
+//
+//  Created by Matt Svoboda on 11/27/23.
+//  Copyright © 2023 lhc. All rights reserved.
+//
+
+import Foundation
+
+class LegacyMigration {
+
+  // Map of [modern pref key → legacy key]. Do not reference legacy keys outside this file.
+  fileprivate static let legacyColorPrefKeyMap: [Preference.Key: Preference.Key] = [
+    Preference.Key.subTextColorString: Preference.Key("subTextColor"),
+    Preference.Key.subBgColorString: Preference.Key("subBgColor"),
+    Preference.Key.subBorderColorString: Preference.Key("subBorderColor"),
+    Preference.Key.subShadowColorString: Preference.Key("subShadowColor"),
+  ]
+
+  /**
+   Loops over the set of legacy preference keys. If a value is found for a given legacy key, but no value is found for its modern equivalent key,
+   the legacy value is migrated & stored under the modern key.
+
+   Older versions of IINA serialized mpv color data into NSObject binary using the now-deprecated `NSUnarchiver` class.
+   This method will transition to the new format which consists of the color components written to a `String`.
+   To do this in a way which does not corrupt the values for older versions of IINA, we'll store the new format under a new `Preference.Key`,
+   and leave the legacy pref entry as-is.
+
+   This method will be executed on each of the affected prefs when IINA starts up. It will first check if there is already an entry for the new
+   pref key. If it finds one, then it will assume that the migration has already occurred, and will just return that.
+   Otherwise it will look for an entry for the legacy pref key. If it finds that, if will convert its value into the new format and store it under
+   the new pref key, and then return that.
+
+   This will have the effect of automatically migrating older versions of IINA into the new format with no loss of data.
+   However, it is worth noting that this migration will only happen once, and afterwards newer versions of IINA will not look at the old pref entry.
+   And since older versions of IINA will only use the old pref entry, users who mix old and new versions of IINA may experience different values
+   for these keys.
+   */
+  static func migrateLegacyPreferences() {
+    let didMigrateColor = Preference.bool(for: .didMigrateColorFromNSArchiver)
```

Comment:
I have a strong opinion on keeping the settings in sync. Normally it would be good to continue support for the old setting, but not in this case due to security. Due to the security issue with `NSArchiver` we want to remove use of the class as soon as practical. So I do not think we should add code to keep the old key in sync. We'd like to eliminate the migration code sooner than later to eliminate all use of `NSArchiver`.

Why not go back to the original scheme where migration occurs when a non-default stored setting for the old key is found and the new key does not have a persisted value? With that scheme no "migrated" flag is needed.

Might want to wait and see what the other reviewers think about this.

Reply:
> Why not go back to the original scheme where migration occurs when a non-default stored setting for the old key is found and the new key does not have a persisted value? With that scheme no "migrated" flag is needed.

The only problem with the previous solution is that the check needs happen before the default values are set, and thus logging is not available. I wouldn't want to wait until after the defaults are set because that would introduce a blind spot for the default value: the user may end up in a state where they want to use the default value, but the migration code will see that as a non-migration and will see their legacy value and will keep re-migrating it every time they launch.

I think the migration flag is a decent solution. The flag can double as a "receipt" for the migration, and while receipts can be annoying, they can sometimes turn out to be pretty useful.

Reply:
On this:
> The only problem with the previous solution is that the check needs happen before the default values are set

That is only required when detecting if a key is persisted based on whether a value is present. Add the following to `Preference` and have `LegacyMigration` use it to check if the modern key is present in the property list.

```swift
  static func keyHasBeenPersisted(key: Key) -> Bool {
    let identifier = InfoDictionary.shared.bundleIdentifier
    guard let persisted = ud.persistentDomain(forName: identifier) else { return false }
    return persisted.keys.contains(key.rawValue)
  }
```